gazebo_plugins/src/gazebo_ros_wheel_slip.cpp

```diff
@@ -256,6 +272,105 @@ void GazeboRosWheelSlip::Load(gazebo::physics::ModelPtr _model, sdf::ElementPtr
       }
     }
   }
+
+  double wheel_spin_tolerance = 1.0e-3;
+  if (!_sdf->HasElement("wheel_spin_tolerance")) {
+    RCLCPP_INFO(
+      impl_->ros_node_->get_logger(), "Missing <wheel_spin_tolerance>, defaults to %f",
+      wheel_spin_tolerance);
+  } else {
+    wheel_spin_tolerance = _sdf->GetElement("wheel_spin_tolerance")->Get<double>();
+  }
+
+  double publisher_update_rate = 100.0;
+  if (!_sdf->HasElement("publisher_update_rate")) {
+    RCLCPP_INFO(
+      impl_->ros_node_->get_logger(),
+      "Missing <publisher_update_rate>, defaults to %f", publisher_update_rate);
+  } else {
+    publisher_update_rate = _sdf->GetElement("publisher_update_rate")->Get<double>();
+  }
+
+  if (publisher_update_rate > 0.0) {
+    impl_->publisher_update_period_ = 1.0 / publisher_update_rate;
+  } else {
+    impl_->publisher_update_period_ = 0.0;
+  }
+
+  impl_->publisher_last_update_time_ = _model->GetWorld()->SimTime();
+
+  impl_->slip_publisher_ = impl_->ros_node_->create_publisher<gazebo_msgs::msg::WheelSlip>(
+    "wheel_slip", 10);
+
+  auto on_update_callback = [this, wheel_spin_tolerance](
+    const gazebo::common::UpdateInfo & info) {
+
+    #ifdef IGN_PROFILER_ENABLE
+      IGN_PROFILE("GazeboRosWheelSlip publisher");
+    #endif
+
+      gazebo::common::Time current_time = info.simTime;
+
+      // If the world is reset, for example
+      if (current_time < impl_->publisher_last_update_time_) {
+        RCLCPP_INFO(impl_->ros_node_->get_logger(), "Negative sim time difference detected.");
+        impl_->publisher_last_update_time_ = current_time;
+      }
+
+      // Check period
+      double seconds_since_last_update = (
+        current_time - impl_->publisher_last_update_time_).Double();
+
+      if (seconds_since_last_update < impl_->publisher_update_period_) {
+        return;
+      }
+
+    #ifdef IGN_PROFILER_ENABLE
+      IGN_PROFILE_BEGIN("fill ROS message");
+    #endif
+
+      // Populate message
+      auto slip_msg = gazebo_msgs::msg::WheelSlip();
+
+      std::map<std::string, ignition::math::Vector3d> slips;
+      this->GetSlips(slips);
+      for (const auto & wheel : slips) {
+        auto name = wheel.first;
+        auto slip = wheel.second;
+        auto spin_speed = slip.Z();
+        auto long_vel = slip.X() + slip.Z();
+        auto lat_vel = slip.Y();
+        double long_slip;
+        if (fabs(spin_speed) < wheel_spin_tolerance) {
+          long_slip = 0;
+        } else {
+          long_slip = -1.0 * (long_vel - spin_speed) / spin_speed;
```

Comment:
nit: this might be easier to read as `long_slip = (spin_speed - long_vel) / spin_speed;`

Reply:
I changed this to be closer to the equation in the pdf you linked above (`long_vel` is in the denominator) and used your suggested styling. The line now reads: `(spin_speed - long_vel) / long_vel`.

Reply:
Actually, I think it makes more sense to have `spin_speed` in the denominator. For example, if the slip is perfect so that the wheel has no longitudinal velocity (i.e., within the tolerance), then the slip would be said to be zero, which it should be 1.0. What do you think? It is a little different than the pdfs equation.

Reply:
there are multiple definitions of wheel slip. During acceleration and at low speeds, it's common to use `spin_speed` in the denominator. At high speeds and during braking, it's common to use `long_vel`. Elsewhere in WheelSlipPlugin we are using the version of slip with `spin_speed` in the denominator, so it may be more consistent to stick with `spin_speed` in the denominator

* https://github.com/osrf/gazebo/blob/gazebo11_11.9.1/plugins/WheelSlipPlugin.cc#L512-L537

Reply:
here's a link to the paper referenced in the WheelSlipPlugin comment:

http://www.astro.mech.tohoku.ac.jp/~yoshida/paperlist/ICRA02-yoshida.pdf